File: content/seadex_api.ts
export type FileEntry = {
    name?: string | null;
    length?: number | null;
};

export type TRSEntry = {
    tracker?: string | null;
    releaseGroup?: string | null;
    url?: string | null;
    dualAudio?: boolean | null;
    isBest?: boolean | null;
    infoHash?: string | null;
    files?: FileEntry[] | null;
};

export type ReleaseData = {
    comparison?: string | null;
    notes?: string | null;
    "theoretical best"?: string | null;
    releases: {
        tracker?: string | null;
        "release group"?: string | null;
        url?: string | null;
        "dual audio"?: boolean | null;
        "is best"?: boolean | null;
        "private tracker"?: boolean | null;
        "file size"?: string | null;
        "episode list"?: {
            name?: string | null;
            size?: string | null;
        }[];
    }[];
};

export class SeadexApi {
    private formatFileSize(bytes: number | null | undefined): string {
        if (bytes == null || isNaN(bytes)) {
            return "Unknown size";
        }

        const megabytes = bytes / (1024 ** 2);
        const gigabytes = bytes / (1024 ** 3);

        if (gigabytes >= 1) {
            return `${gigabytes.toFixed(1)} GiB`;
        } else {
            return `${megabytes.toFixed(1)} MiB`;
        }
    }

    async getReleaseData(anilistId: number): Promise<ReleaseData | null> {
        const baseUrl = "https://releases.moe/api/collections/entries/records";
        const trsUrl = `${baseUrl}?filter=alID=${anilistId}&expand=trs`;

        const response = await fetch(trsUrl);
        if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
        }

        const data = await response.json();
        const items = data?.items?.[0] || null;
        if (!items) return null;

        const releaseDict: ReleaseData = {
            comparison: items.comparison ?? null,
            notes: items.notes ?? null,
            "theoretical best": items.theoreticalBest ?? null,
            releases: []
        };

        const trs = items.expand?.trs || [];
        for (const entry of trs as TRSEntry[]) {
            let totalFileSize = 0;

            const episodeList =
                entry.files?.map(file => {
                    const size = this.formatFileSize(file.length);
                    if (file.length) totalFileSize += file.length;
                    return {
                        name: file.name ?? "Unknown",
                        size
                    };
                }) ?? [];

            const totalFileSizeFormat = this.formatFileSize(totalFileSize);
            const privateTracker = entry.infoHash === "<redacted>";

            releaseDict.releases.push({
                tracker: entry.tracker ?? "Unknown",
                "release group": entry.releaseGroup ?? "Unknown",
                url: entry.url ?? "",
                "dual audio": entry.dualAudio ?? false,
                "is best": entry.isBest ?? false,
                "private tracker": privateTracker,
                "file size": totalFileSizeFormat,
                "episode list": episodeList
            });
        }

        return releaseDict;
    }
}

// const api = new SeadexApi();
// api.getReleaseData(104276)
//     .then(data => {
//         console.log(JSON.stringify(data, null, 2));
//     })
//     .catch(err => console.error(err));

File: py-archive/seadex_api.py
import requests


class SeadexApi:
    def get_release_data(self, anilist_id):
        base_url = "https://releases.moe/api/collections/entries/records"
        trs_url = f"{base_url}?filter=alID={anilist_id}&expand=trs"
        response = requests.get(trs_url)
        response.raise_for_status()

        data = response.json()

        items = data.get("items")
        items = items[0] if items else None

        comparison = items.get("comparison")
        notes = items.get("notes")
        theoretical_best = items.get("theoreticalBest")

        release_dict = {
            "comparison": comparison,
            "notes": notes,
            "theoretical best": theoretical_best,
            "releases": [],
        }

        trs = items.get("expand")
        trs = trs.get("trs")

        for entry in trs:
            tracker = entry.get("tracker")
            release_group = entry.get("releaseGroup")
            url = entry.get("url")
            dual_audio = entry.get("dualAudio")
            is_best = entry.get("isBest")

            files = entry.get("files")
            total_file_size = 0

            episode_list = []

            for file in files:
                name = file.get("name")
                file_size_format = self.format_file_size(file.get("length"))
                total_file_size += file.get("length", 0)

                episode_list.append({"name": name, "size": file_size_format})

            total_file_size_format = self.format_file_size(total_file_size)

            if entry.get("infoHash") == "<redacted>":
                private_tracker = True
            else:
                private_tracker = False

            entry_dict = {
                "tracker": tracker,
                "release group": release_group,
                "url": url,
                "dual audio": dual_audio,
                "is best": is_best,
                "private tracker": private_tracker,
                "file size": total_file_size_format,
                "episode list": episode_list,
            }

            release_dict["releases"].append(entry_dict)

        return release_dict

    def format_file_size(self, bytes):
        megabytes = bytes / (1024**2)
        gigabytes = bytes / (1024**3)

        if gigabytes >= 1:
            return f"{round(gigabytes, 1)}" + " GiB"
        else:
            return f"{round(megabytes, 1)}" + " MiB"


if __name__ == "__main__":
    print(SeadexApi().get_release_data(104276))
